### PR TITLE
GEODE-5502: Removed duplicate / member-specific receivers from cluste…

### PR DESCRIPTION
**Pull request by boglesby**
…r configuration

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/test/java/org/apache/geode/distributed/internal/InternalConfigurationPersistenceServiceTest.java [hunk 1]
```diff
@@ -193,4 +204,93 @@ public void getPackagesToScanWithSystemProperty() {
...
+  private String getValidReceiversXml() {
+    return "<cache>\n<gateway-receiver/>\n</cache>";
+  }
+
+  protected Object[] getXmlAndExpectedElements() {
+    return $(
+        new Object[] {getDuplicateReceiversWithDefaultPropertiesXml(), 2, 1},
+        new Object[] {getDuplicateReceiversWithDifferentHostNameForSendersXml(), 2, 0},
+        new Object[] {getDuplicateReceiversWithDifferentBindAddressesXml(), 2, 0},
+        new Object[] {getValidReceiversXml(), 1, 1});
```
[the remark this answers is not in the document]
Comment (boglesby) [reply geode-core/src/test/java/org/apache/geode/distributed/internal/InternalConfigurationPersistenceServiceTest.java]: Good idea. I'll add one.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/InternalConfigurationPersistenceService.java [hunk 2]
```diff
@@ -543,6 +544,65 @@ void initSharedConfiguration(boolean loadSharedConfigFromDir)
...
+          configuration.setCacheXmlContent(XmlUtils.prettyXml(document));
+          configRegion.put(group, configuration);
+        }
+      }
+    }
+  }
+
+  boolean removeInvalidGatewayReceivers(Document document) throws TransformerException {
+    boolean modified = false;
+    NodeList receiverNodes = document.getElementsByTagName("gateway-receiver");
```
Comment (mcmellawatt) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/InternalConfigurationPersistenceService.java]: I'm not sure on our stance on how much input validation to do, but if the malformed XML doesn't have the "gateway-receiver" element then I think we would get an NPE on the next line when we call receiverNodes.getLength().  Maybe its reasonable to expect a customer wouldn't accidentally remove this element though.

Comment (boglesby) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/InternalConfigurationPersistenceService.java]: Thanks. Let me add a test for that case. Its certainly possible that there are no receivers in the xml.

Comment (boglesby) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/InternalConfigurationPersistenceService.java]: It works ok with no receivers in the xml:

Initial xml content:
```
<cache>
</cache>
```
Processed xml content:
```
<cache>
</cache>
```

**Pull request merged**